### PR TITLE
Line numbers and inline editing

### PR DESCRIPTION
**Pull request by CatoTH**
New features:
- Line numbers in the motion detail view (including projector view). The maximum line length is configurable.
- Inline editing of motion text (if the "update"-permission is present and the most recent version of the motion is selected)
- Unit-tests based on Karma

The functions in diff.js are in preparation for the next feature (amendments)

### CODE REVIEW

File: openslides/core/static/css/projector.css [hunk 1]
```diff
@@ -335,3 +335,51 @@ tr.elected td {
...
+    content: attr(data-line-number);
+    vertical-align: top;
+    font-size: 11px;
+    color: gray;
+    font-family: Courier, serif;
+    margin-top: -3px;
+    margin-left: 0;
+    margin-right: 3px;
+}
+
```
Comment (normanjaeckel) [inline openslides/core/static/css/projector.css]: Remove this line.

File: openslides/motions/static/js/motions/diff.js [hunk 2]
```diff
@@ -0,0 +1,413 @@
+(function () {
+
+"use strict";
+
+angular.module('OpenSlidesApp.motions.diff', ['OpenSlidesApp.motions.lineNumbering'])
+
+/**
+ * TO DO
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/diff.js]: Can this TODO be removed?

File: openslides/motions/static/js/motions/diff.js [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        var html = '<' + node.nodeName;
+        for (var i = 0; i < node.attributes.length; i++) {
+          var attr = node.attributes[i];
+          html += " " + attr.name + "=\"" + attr.value + "\"";
+        }
+        html += '>';
+        return html;
+    };
+
+    this._serializeDom = function(node, stripLineNumbers) {
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/diff.js]: Can you please add some comments about the return statements inside the fucntion (why early return).

File: openslides/motions/static/js/motions/linenumbering.js [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    this._isInlineElement = function (node) {
+        var inlineElements = [
+            'SPAN', 'A', 'EM', 'S', 'B', 'I', 'STRONG', 'U', 'BIG', 'SMALL', 'SUB', 'SUP', 'TT'
+        ];
+        return (inlineElements.indexOf(node.nodeName) > -1);
+    };
+
+    this._isOsLineBreakNode = function (node) {
+        if (!node) {
+            return false;
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/linenumbering.js]: Here the same.

File: openslides/motions/static/js/motions/site.js [hunk 5]
```diff
@@ -651,6 +667,84 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions'])
                     $scope.version = motion.active_version;
                 });
         };
+
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Remove one blank line.

File: openslides/motions/static/js/motions/site.js [hunk 6]
```diff
@@ -651,6 +667,84 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions'])
...
+        };
+
+        if (motion.state.versioning && Config.get('motions_allow_disable_versioning').value) {
+            $scope.inlineEditing.trivialChange = true;
+            $scope.inlineEditing.trivialChangeAllowed = true;
+        }
+
+        $scope.tinymceOptions = Editor.getOptions(null, true);
+        $scope.tinymceOptions.readonly = 1;
+        $scope.tinymceOptions.entities = "160,nbsp,38,amp,34,quot,162,cent,8364,euro,163,pound,165,yen,169,copy," +
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Please add a short note about this magic numbers.

File: openslides/motions/static/js/motions/site.js [hunk 7]
```diff
@@ -651,6 +667,84 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions'])
...
+            "174,reg,8482,trade,8240,permil,60,lt,62,gt,8804,le,8805,ge,176,deg,8722,minus";
+        $scope.tinymceOptions.setup = function(editor) {
+            $scope.inlineEditing.editor = editor;
+            editor.on("change", function() {
+                var text = normalizeInlineHtml(editor.getContent());
+                text = text.replace(/ \/>/g, ">");
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Please add a short note what this reges is doing.

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 8]
```diff
@@ -286,9 +339,11 @@ <h3 translate>Version history</h3>
...
               <small>{{ message.message }}</small>
+            </li>
+          </ul>
         </div>
       </div>
     </div>
 
   </div>
-</div>
+</div>
```
Comment (normanjaeckel) [inline openslides/motions/static/templates/motions/motion-detail.html]: Please add linebreak line at the end.

File: package.json [hunk 9]
```diff
@@ -20,6 +20,12 @@
     "po2json": "~0.4.1",
     "sprintf-js": "~1.0.3",
     "through2": "~2.0.0",
-    "yargs": "~3.32.0"
+    "yargs": "~3.32.0",
+
+    "karma": "~1.1.0",
```
Comment (normanjaeckel) [inline package.json]: The new deps can be sorted into the current ...

**Comment by emanuelschuetze** [comment]
Tested
* outside numbering on projector
* hide save button on first opening
* blockquote css font size
* title for linenumbering

File: openslides/motions/static/js/motions/site.js [hunk 10]
```diff
@@ -932,22 +936,34 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions', 'OpenSlid
...
-        $scope.tinymceOptions.entities = "160,nbsp,38,amp,34,quot,162,cent,8364,euro,163,pound,165,yen,169,copy," +
-            "174,reg,8482,trade,8240,permil,60,lt,62,gt,8804,le,8805,ge,176,deg,8722,minus";
-        $scope.tinymceOptions.setup = function(editor) {
+        $scope.tinymceOptions.setup = function (editor) {
             $scope.inlineEditing.editor = editor;
-            editor.on("change", function() {
-                var text = normalizeInlineHtml(editor.getContent());
-                text = text.replace(/ \/>/g, ">"); // Removes slashes in self-closing tags, like <IMG /> -> <IMG>
-                $scope.inlineEditing.changed = (text != $scope.inlineEditing.originalHtmlNormalized);
+
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Remove blank line.

File: openslides/motions/static/js/motions/site.js [hunk 11]
```diff
@@ -932,22 +936,34 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions', 'OpenSlid
...
 
         $scope.enableInlineEditing = function() {
             $scope.inlineEditing.editor.setMode("design");
             $scope.inlineEditing.active = true;
+            $scope.inlineEditing.changed = false;
+
+            $scope.lineBrokenText = motion.getTextWithLineBreaks($scope.version);
+            $scope.inlineEditing.editor.setContent($scope.lineBrokenText);
+            $scope.inlineEditing.originalHtml = $scope.inlineEditing.editor.getContent();
+            window.setTimeout(function() {
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Should we use angular `$timeout` here?

**Comment by emanuelschuetze** [comment]
Last two wording/styling issues:
* Please use always "inline" (instead of "inside") in all ui textes, comments, variables, css-classes (currently there is a mix of both)
* motion-detail-view: reduce the space between line numbers and text in "outside" numbering view (space in projector motion slide looks good!)

Then we can merge...

**Pull request closed without merging**